### ISSUE TITLE
Prematurely resolved submission of Apple Pay Component

### ISSUE THREAD

**Issue by veewee** [Adyen/adyen-web#1926]
**Describe the bug**

We are using Apple Pay through the Adyen web components.
In big lines, our code follows the documentation https://docs.adyen.com/payment-methods/apple-pay/web-component

The main difference is that we are not using the apple pay button. (`showPayButton: false`).
In our system, the customer can select a payment method and later on click on one main "complete order" submit button.
At that moment, we start the payment flow like this:

```js
await element.submit();
const thePaymentData = element.data;
```

We pass this data to our own endpoint, which sends it to Adyen's /payments API.
*Note:* We rely on the component to provide the data, not on the checkout's onPaymentCompleted option.

The problem with this, is that there is a subtle "bug" in your apple pay component:
https://github.com/Adyen/adyen-web/blob/70b6e12cf810e622975e548e67d99d6cccaea0b0/packages/lib/src/components/ApplePay/ApplePay.tsx#L60-L88
 
 
How it works now:

* Submit() triggers startSession() which returns a promise.
* This promise can be accepted/rejected with the onClick callback
* Based on this decision the apple session logic is started.

This is where it goes wrong: the promise resolves instantly without keeping track of the session.
So if you:

```js
await element.submit()
```

The promise gets resolved instantly before the payment popup is even opened.
This results in us sending the payment date (without approved apple pay token) to your /payments endpoint, which results in an error.
You're basically telling us the payment has completed even though it hasn't started yet.

**Expected behavior**

From a developer point of view, I'dd expect the submit only be resolved once the complete payment flow has been finished.
It works like this in e.g. [the google pay component](https://github.com/Adyen/adyen-web/blob/70b6e12cf810e622975e548e67d99d6cccaea0b0/packages/lib/src/components/GooglePay/GooglePay.tsx#L60).

This can be achieved by letting the promise that instantly returns void, instead return a promise that watches the onCancel (for reject) and the onPaymentAuthorized (for resolving) in order to complete the submition of the form.

An **untested** implementation could look like this:

```js
return new Promise((resolve, reject) => this.props.onClick(resolve, reject)).then(() => {
    return new Promise((submitFinished, submitFailed) => {
        const paymentRequest = preparePaymentRequest({
            companyName: this.props.configuration.merchantName,
            ...this.props
        });

        const session = new ApplePayService(paymentRequest, {
            version,
            onCancel: (reason) => {
                onCancel(reason);
                submitFailed(reason);
            },
            onPaymentMethodSelected,
            onShippingMethodSelected,
            onShippingContactSelected,
            onValidateMerchant: onValidateMerchant || this.validateMerchant,
            onPaymentAuthorized: (resolve, reject, event) => {
                if (!!event.payment.token && !!event.payment.token.paymentData) {
                    this.setState({ applePayToken: btoa(JSON.stringify(event.payment.token.paymentData)) });
                }

                super.submit();
                onPaymentAuthorized(resolve, reject, event);
                submitFinished(event);
            }
        });

        session.begin();
    });
})
```

❗  all listeners could be rejecting the promise here. Eg `onValidateMerchant` or `validateMerchant`, `onPaymentMethodSelected`, `onShippingMethodSelected`, `onShippingContactSelected` could fail as well, which must result in a rejection of the main promise. Currently there is e.g. no way to handle a failure of the internal `validateMerchant` function from outside the component or reuse that function from the outside.

**"Crappy" workaround in our application**

To overcome this problem, we had make sure to check the outcome of 3 functions: submit(), onAuthorized callback and onCancel callback.
Even though it works, it is not what you would like to go through from an implementators point of view.

For reference, here is our workaround:

```ts
function useApplePaySpecificSubmitLogic()
{
    const paymentAttempt = useRef(new Deferred<ApplePayJS.ApplePayPaymentAuthorizedEvent>());

    // Await on authorized event
    const onAuthorized = (resolve: any, reject: any, event : ApplePayJS.ApplePayPaymentAuthorizedEvent) => {
        console.log('authorized', event);
        paymentAttempt.current.resolve(event);
        resolve();
    }

    // Bail out on error event
    const onError = (reason: any) => {
        console.log('error', reason);
        paymentAttempt.current.reject(reason);
    }

    // A factory for creating the build logic based on an element.
    const createSubmitLogicForElement = function createsubmitLogic(element: UIElement) {
        // Starts the element submit() and then waits for the authorization to kick in...
        return async () => {
            paymentAttempt.current = new Deferred();

            // This finishes way to early (after onClick)
            await element.submit();

            // This waits for onAuthorized or onError
            await paymentAttempt.current.promise;

            return element.data
        }
    }

    return {
        createSubmitLogicForElement,
        onAuthorized,
        onError
    }
}
```

```ts
const {onAuthorized, onError, createSubmitLogicForElement} = useApplePaySpecificSubmitLogic();
const element = adyenCheckout
    .create('applepay', {
        ...options,
        onAuthorized,
        onError
    } as ApplePayElementProps);

element
    .isAvailable()
    .then(
        () => {
            element.mount(container);
            onPaymentSelected(createSubmitLogicForElement(element));
        }
    );

```

It would be nice to see this fixed in Adyen's core web components so that the submit() promise could be used to await payment completion.

Thanks!

**Comment by jonathanribas** [comment]
Hi, we are experiencing same issue too.
@ribeiroguilherme any chance to have priority on this one please?

**Comment by ribeiroguilherme** [comment]
Hi @veewee ,

Thanks for the well explained issue 👍  I am taking a look into it and will keep this post updated.

May I ask you: 
- Are you using the `onSubmit` callback for ApplePay? How are you implementing this? I'd like to understand this a bit better, as inside the `onPaymentAuthorized` we eventually trigger the `onSubmit` flow, but seems like you are not using it.
- I see in your proposal that `submitFinished` resolves passing the apple pay `event` - do you see any use case for this? We already provide the event for the `onAuthorised` , so perhaps the Promise can simply be resolved without passing any value to it

**Comment by veewee** [comment]
Hello,

* No we are not using the `onSubmit` callback currently. To give a bit more context on why: Adyen is not the only payment method available in our application. Therefore we want to use adyen's checkout library, but avoid sending everything through it. We render all adyen components we want the customers to use and let these components provide the payment details to the main (react) application. This will then decide what to send to our server, finish the order, ... This way, we can use Adyen's checkout library as some way of paying, without being inside the core part of the application logic.

* `submitFinished` indeed passes the apple pay event. I just made this consistent with the google-pay component. I'm not using the event myself - but I can imagine it makes sense that your return the successful event object instead of just void.



One other thing I noticed in the implementation above:
Since we do not have a sandbox environment for apple pay in Belgium, trying to use apple pay results in an error on the test environment. 
This happens in `this.validateMerchant` that is being triggered in the `startSession`:

https://github.com/Adyen/adyen-web/blob/master/packages/lib/src/components/ApplePay/ApplePay.tsx#L75

With the workaround code pasted above, the payment will never get resolved nor rejected on the invalid test environment, which freezes our pay button. If you are reworking the thing, you might want to reject the main promise if this callback fails as well.

**Comment by ribeiroguilherme** [comment]
> ❗ all listeners could be rejecting the promise here. Eg onValidateMerchant or validateMerchant, onPaymentMethodSelected, onShippingMethodSelected, onShippingContactSelected could fail as well, which must result in a rejection of the main promise. Currently there is e.g. no way to handle a failure of the internal validateMerchant function from outside the component or reuse that function from the outside.

The `this.validateMerchant` should reject the submit Promise indeed.

I am not 100% sure about the other callbacks (`onPaymentMethodSelected, onShippingMethodSelected, onShippingContactSelected`):  The `reject` parameter can be called on those methods, but that doesn't mean that the submit flow should fail - perhaps the merchant wants to reject the option that the shopper chose in the ApplePay UI, and then the shopper can opt for something else. In such scenario, the main submit Promise shouldn't be rejected as the shopper can still continue the payment flow.

**Comment by veewee** [comment]
Thanks for the reply.

I'll have to play around with this information from our side as well.

The reason we went for this approach is because we had Google pay before Apple Pay and continued from that code.
Since our project has promise based resolutions, it would still require us to wrap a deferred around this `submit()` + wait for `onSubmit` or `onError` callback.

That is the reason why I think for situations like these, having a full promise solution for both apple pay and google pay (and maybe others) probably results in a better developer experience: it allows people to implement Adyen checkout with both callbacks and promises - making it less blocking on those systems.

**Comment by ribeiroguilherme** [comment]
I totally understand your expectations here and I see that from the dev experience perspective, it can help how to handle the submit flow. Anyway, I took your suggestion into consideration and I added to our improvements ideas for the next major release which we are planning to work on it soon. Perhaps this could be something that we can investigate/add on it.

I will close this issue for now as there is nothing to be done on our side currently. Thanks for reaching out!
